File: 6_grupos/retrovisor.py
import re
string = '<b>Destaque</b><strong>Forte</strong><div>Conteudo</div>'

print(re.findall('<(\w+)>', string))    # Não inclui '< >'
# ['b', 'strong', 'div']

print(re.findall('(<\w+>)', string))    # Inclui '< >'
# ['<b>', '<strong>', '<div>']

print(re.findall(r'<(\w+)>.*<\/\1>', string))    # Retrovisores, \1 representa o resultado do grupo 1
# ['b', 'strong', 'div']

string2 = 'Lentamente é mente muito lenta.'
result = re.findall(r'(lenta)(mente)', string2, re.I)
print(result)
# [('Lenta', 'mente')]
print(''.join(list(result)[0]))
# Lentamente

result2 = re.findall(r'(lenta)(mente).*\2.*\1', string2, re.I)
print(result2)
# [('Lenta', 'mente')]

File: 6_grupos/retrovisor.js
string = '<b>Destaque</b><strong>Forte</strong><div>Conteudo</div>'

console.log(string.match(/<(\w)>.*<\/\1>/g))
// [ '<b>Destaque</b>' ]

string = 'Lentamente é mente muito lenta.'
console.log(string.match(/(lenta)(mente)/gi))
// [ 'Lentamente' ]

console.log(string.match(/(lenta)(mente).*\2/gi))
// [ 'Lentamente é mente' ]

console.log(string.match(/(lenta)(mente).*\2.*\1\./gi)) // retrovisor 1 = lenta, retrovisor 2 = mente
// [ 'Lentamente é mente muito lenta' ]

console.log(string.match(/(?:lenta)(mente).*\1/gi)) // ?: não armazena o valor, retrovisor 1 é mente
// [ 'Lentamente é mente' ]

console.log(string.match(/(lenta)(mente)?/gi)) // mente é opcional
// [ 'Lentamente', 'lenta' ]

console.log(string.replace(/(lenta)(mente)/gi, '$2')) // $2 é o retorvisor 'mente', replace subistitui 'Lentamente' por 'mente'
// mente é mente muito lenta.

console.log(string.replace(/(lenta)(mente)/gi, '$2 lenta'))
// mente lenta é mente muito lenta.


const texto3 = 'abcdefghijkll'
console.log(texto3.match(/(a)(b)(c)(d)(e)(f)(g)(h)(i)(j)(k)(l)\12/g))